### DISCUSSION TITLE
An alternative approach to soft deletion using hooks

### DISCUSSION THREAD

**Discussion by kisamoto** [pocketbase/pocketbase#2694] [Show and tell]
Sometimes it's useful to 'soft delete', that is to act as if the record has been deleted however keep it for audit purposes. Commonly this is done by setting a `deleted: true` or `deletedAt: 2004-06-14T23:34:30` attribute on the row however you then need to ensure you filter these rows out in your security rules or API responses.

Alternatively we can actually delete the row from the table however maintain a copy in a flexible JSON structure in a separate table.

There is [an excellent blog post](https://brandur.org/fragments/deleted-record-insert) that goes into more details for this approach and on how to achieve this with SQL functions (e.g. with PostgreSQL).

SQLite doesn't seem to support standard SQL functions however we can also leverage the power of PocketBase hooks to the same effect.

Below is the code for the `BeforeDelete` hook that I use. I chose `BeforeDelete` so that if this fails for whatever reason I don't lose data (as if using `AfterDelete` the record may be removed but if this hook fails we may not have a record of it. Adapt to your needs). 

You need to create a dedicated collection (I called mine `deleted`), which is only accessible via admins and has the following two columns (additional to the `id`, `created`, `updated`):

Column name | Column type | Purpose
--- | ---- | ----
`collection` | Plain text | Extract the column name explicitly for querying
`record` | JSON | A JSON representation of the deleted row

When deleting, you'll end up with something like the following (example from `users` collection. Note: No passwords are included):

![image](https://github.com/pocketbase/pocketbase/assets/1744908/0d6448f2-e441-46b2-a3f2-24b3ead0d895)

```go

// Used for the `slices.Contains` utility function
import "golang.org/x/exp/slices"

// Helper hook that keeps a record of the deleted data as an alternative to soft deletes
// Inspiration: https://brandur.org/fragments/deleted-record-insert
app.OnRecordBeforeDeleteRequest().Add(func(e *core.RecordDeleteEvent) error {
  const (
	  DeletedCollectionName = "deleted"
  )
  
  // Add other collection names here where you don't want to keep a copy of the deleted record
  excludedCollections := []string{
	  DeletedCollectionName,
  }
  
  // Skip if the record is already deleted or in excluded collections
  if slices.Contains(excludedCollections, e.Collection.Name) {
	  return nil
  }
  
  collection, err := app.Dao().FindCollectionByNameOrId(DeletedCollectionName)
  if err != nil {
	  return err
  }
  
  record := models.NewRecord(collection)
  
  form := forms.NewRecordUpsert(app, record)
  form.LoadData(map[string]any{
	  "collection": e.Collection.Name,
	  "record":     e.Record,
  })
  
  return form.Submit()
})
```

Let me know what you think and if/how you would improve it.

**Comment by HEAVYPOLY** [comment]
I want a feature like this and ended up just adding deletedDate in my collection but this looks much cleaner especially in admin UI!

**Reply by kisamoto** [reply]
It also has the added benefit that you don't always have to remember to filter using `deletedDate` so can also make your queries a little safer.

**Comment by Vano2903** [comment]
i don't know if there are newer/preferred solutions to this right now but I updated the code to work on newer versions (currently tested on the v0.24.3)

```go
app.OnRecordDelete().BindFunc(func(e *core.RecordEvent) error {
    l := app.Logger()

    const (
        DeletedCollectionName = "deleted"
    )

    // Add other collection names here where you don't want to keep a copy of the deleted record
    excludedCollections := []string{
        DeletedCollectionName,
    }

    recordCollection := e.Record.Collection().Name

    // Skip if the record is already deleted or in excluded collections
    if slices.Contains(excludedCollections, recordCollection) {
        l.Debug("Skipping soft delete", "record", e.Record, "collection", recordCollection)
        return e.Next()
    }

    collection, err := app.FindCollectionByNameOrId(DeletedCollectionName)
    if err != nil {
        l.Error("Failed to find collection", "collection", DeletedCollectionName, "error", err)
        return err
    }

    record := core.NewRecord(collection)

    form := forms.NewRecordUpsert(app, record)
    form.Load(map[string]any{
        "collection": recordCollection,
        "record":     e.Record,
    })

    app.Logger().Info("Soft deleting record", "record", e.Record, "collection", recordCollection)

    if err := form.Submit(); err != nil {
        l.Error("Failed to soft delete record", "record", e.Record, "collection", recordCollection, "error", err)
        return err
    }

    return e.Next()
})
```

We could even add a bootstrap hook to check if the deleted collection exists I think it's just extra work that it's not needed because we cannot control the settings of the soft deletion using the interface

**Reply by ganigeorgiev** [reply]
_Side-note: there is no need to use `forms.NewRecordUpsert` anymore with v0.23+. Please read the upgrade guide in https://pocketbase.io/v023upgrade/go/._

**Reply by Vano2903** [reply]
Thanks for sharing this document! i updated the code to use the record using the .Set function
```go
app.OnRecordDelete().BindFunc(func(e *core.RecordEvent) error {
    l := e.App.Logger()

    const (
        DeletedCollectionName = "deleted"
    )

    // Add other collection names here where you don't want to keep a copy of the deleted record
    excludedCollections := []string{
        DeletedCollectionName,
    }

    recordCollection := e.Record.Collection().Name

    // Skip if the record is already deleted or in excluded collections
    if slices.Contains(excludedCollections, recordCollection) {
        l.Debug("Skipping soft delete", "record", e.Record, "collection", recordCollection)
        return e.Next()
    }

    deleteCollection, err := e.App.FindCollectionByNameOrId(DeletedCollectionName)
    if err != nil {
        l.Error("Failed to find collection", "collection", DeletedCollectionName, "error", err)
        return err
    }

    newDeletedRecord := core.NewRecord(deleteCollection)

    newDeletedRecord.Set("collection", recordCollection)
    newDeletedRecord.Set("record", e.Record)

    if err := e.App.Save(newDeletedRecord); err != nil {
        l.Error("Failed to soft delete record", "record", e.Record, "collection", recordCollection, "error", err)
        return err
    }

    return e.Next()
})
```

**Reply by ganigeorgiev** [reply]
_Also if you have to access the app instance from within a Go hook, prefer using the `e.App` field because with v0.23+ we allow wrapping the entire execution chain in a transaction (see the note in https://pocketbase.io/docs/go-event-hooks/)_

**Reply by Vano2903** [reply]
thanks, i edited the code above to use the e.App instead, if there are more issues/misuses, please let me know

**Reply by SlowLogicBoy** [reply]
I like `isDeleted` and `dateDeleted` columns so I use this:
```go
app.OnRecordDelete().BindFunc((func(e *core.RecordEvent) error {
    isDeletedField := e.Record.Collection().Fields.GetByName("isDeleted")
    //Skip if does not support soft delete
    if isDeletedField == nil {
        return e.Next()
    }
    //Hard delete
    isDeleted := e.Record.GetBool("isDeleted")
    dateDeleted := e.Record.GetDateTime("dateDeleted")
    if isDeleted {
        if dateDeleted.Before(types.NowDateTime().AddDate(0, 0, -90)) { //TODO: configurable (currently delete after 90 days
            return e.Next()
        }
        //Do nothin
        return nil
    }
    //Soft delete
    e.Record.Set("isDeleted", true)
    e.Record.Set("dateDeleted", types.NowDateTime())
    if err := e.App.Save(e.Record); err != nil {
        return err
    }
    return nil
}))
```
I'm not good with go and don't have much knowledge on pocketbase api so...
I am sure I am causing some issues by not calling `e.next()`.
~~The only slight problem is I guess listing records `!isDeleted` but I'm sure I can do those with pocketbase list rules.~~
With:
![image](https://github.com/user-attachments/assets/680ac97b-0add-4918-8a6f-9a2aaac20e84)

**Comment by tewshi** [comment]
a version of what I use at the moment:

```go
// soft delete hook, for the first time, set the deleted field to the current timestamp
app.OnRecordDelete().BindFunc(func(e *core.RecordEvent) error {
	isDeletedField := e.Record.Collection().Fields.GetByName("deleted")
	// skip if collection does not support soft delete
	if isDeletedField == nil || isDeletedField.Type() != core.FieldTypeDate {
		return e.Next()
	}

	dateDeleted := e.Record.GetString("deleted")

	if dateDeleted != "" {
		// if record was soft deleted, delete it permanently
		return e.Next()
	}

	e.Record.Set("deleted", time.Now().UTC())

	// save without triggering record update hook
	e.App.UnsafeWithoutHooks().SaveNoValidate(e.Record)

    // todo: probably notify real-time listeners

	// trigger record deleted success hook
	return e.App.UnsafeWithoutHooks().OnRecordAfterDeleteSuccess(e.Record.Collection().Name).Trigger(e)
})
```

wish I can have something similar to filter only items that have empty `deleted` fields, in one place (instead of using rules for every collection, one can forget when creating a new collection)

**Comment by lekv** [comment]
I'm trying to implement a soft-delete pattern with a deleted_at column at the moment, too. For that I also added `deleted_at = null` to the API rules so that normal client operations don't have to worry about seeing the deleted records. However, I also use a realtime subscription to track changes to records on the client side. This worked well so far but with soft-deletes it now breaks because once `deleted_at` is set, pocketbase doesn't send out the event for the soft-deleted record and I miss it on the client side. Are there any recommended workarounds or patterns for soft-delete with realtime updates? Thanks for the help :)
